### PR TITLE
Fix "Host not found" error

### LINKED ISSUE #147: "Host not found" error stack for users not allowed to see a host/service node details
Our helpdesk has an account in icingaweb2 and can access a specific Business Process in order to check that business applications are available.
It works well, however they get an error when clicking on a host/service after going through all the layers of a particular service. See the screenshot.
The error is due to the fact that this helpdesk user is not allowed to view hosts / services.
Instead of a 'Host not found' error stack, I would expect a 'Not allowed action' error well rendered.

## Expected Behavior

'Not allowed action' error well rendered (not a stack).

## Current Behavior

'Host not found' error stack
```
#0 /usr/share/php/Icinga/Exception/IcingaException.php(41): ReflectionClass->newInstanceArgs(Array)
#1 /usr/share/php/Icinga/Web/Controller.php(78): Icinga\Exception\IcingaException::create(Array)
#2 /usr/share/icingaweb2/modules/monitoring/application/controllers/HostController.php(33): Icinga\Web\Controller->httpNotFound('Host not found')
#3 /usr/share/php/Icinga/Web/Controller/ActionController.php(139): Icinga\Module\Monitoring\Controllers\HostController->init()
#4 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(59): Icinga\Web\Controller\ActionController->__construct(Object(Icinga\Web\Request), Object(Icinga\Web\Response), Array)
#5 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#6 /usr/share/php/Icinga/Application/Web.php(389): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#7 /usr/share/php/Icinga/Application/webrouter.php(109): Icinga\Application\Web->dispatch()
#8 /usr/share/icingaweb2/public/index.php(4): require_once('/usr/share/php/...')
#9 {main}
```
## Steps to Reproduce (for bugs)

1. create an "helpdesk" icingaweb2 user with not host/service object permissions
2. as admin, create business process with a process node and a related host node
3. allow the helpdesk user to view this business process
4. as helpdesk user, go through the business process -> process node -> host node. The host node details will produce this error

## Your Environment

* Module version (System - About): businessprocess 2.1.0
* Icinga Web 2 version and modules (System - About): IcingaWeb2 2.4.2
* Icinga 2 version (`icinga2 --version`):
* Operating System and version:
* Webserver, PHP versions:

### PR DESCRIPTION
**Pull request by Al2Klimov**
fixes #147

### CODE REVIEW

File: library/Businessprocess/HostNode.php [hunk 1]
```diff
@@ -64,7 +65,20 @@ public function getUrl()
...
+        } else {
+            $query = $this->createBackend()->select()
+                ->from('hoststatus', array('host_name'))
+                ->where('host_name', $this->hostname);
+
+            $accessible = Monitoring::restrict($query)->fetchRow() !== false;
+        }
+
+        return Url::fromPath(
+            $accessible ? 'monitoring/host/show' : 'businessprocess/monitored-object/access-denied',
```
Comment (lazyfrosch) [inline library/Businessprocess/HostNode.php]: Where does the `access-denied` URL lead?

File: library/Businessprocess/Monitoring.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Factory
+     *
+     * @return self
+     */
+    final private static function getInstance()
+    {
+        if (static::$instance === null) {
+            static::$instance = new self(Icinga::app()->getRequest(), Icinga::app()->getResponse());
```
Comment (lazyfrosch) [inline library/Businessprocess/Monitoring.php]: This seems like a pretty ugly hack

File: library/Businessprocess/ServiceNode.php [hunk 3]
```diff
@@ -51,6 +52,20 @@ public function getUrl()
...
+            $query = $this->createBackend()->select()
+                ->from('servicestatus', array('host_name', 'service_description'))
+                ->where('host_name', $this->hostname)
+                ->where('service_description', $this->service);
+
+            $accessible = Monitoring::restrict($query)->fetchRow() !== false;
+        }
+
+        return Url::fromPath(
+            $accessible ? 'monitoring/service/show' : 'businessprocess/monitored-object/access-denied',
```
Comment (lazyfrosch) [inline library/Businessprocess/ServiceNode.php]: Where does the `access-denied` URL lead?

File: library/Businessprocess/Monitoring.php (hunk 2 above)
Comment (Al2Klimov) [reply library/Businessprocess/Monitoring.php]: I can't use arbitrary info from the monitoring backend.
I have to respect the current user's restrictions – with the code provided by the current IW2 release.

Comment (lazyfrosch) [reply library/Businessprocess/Monitoring.php]: I just wished there would be a better way... *sigh* :wink:

File: library/Businessprocess/HostNode.php (hunk 1 above)
Comment (Al2Klimov) [reply library/Businessprocess/HostNode.php]: It seems I've forgotten to add some file to Git.. 😕

**Comment by Thomas-Gelf** [comment]
As discussed with @lazyfrosch, for performance reasons this is not the right way of fixing this. Please make all links point to a page in this module and let that controller check accessibility. If fine, redirect to the object. If not, show a friendly hint.

**Pull request merged**